src/storage.rs

```diff
@@ -14,99 +16,333 @@ pub trait StorageBase {
+    type UnprotectedStorage = InnerVec<T>;
+
     fn new() -> Self {
-        let fnv = BuildHasherDefault::<FnvHasher>::default();
-        HashMapStorage(HashMap::with_hasher(fnv))
+        VecStorage {
+            set: BitSet::new(),
+            values: InnerVec(Vec::new()),
+        }
     }
-    fn get(&self, entity: Entity) -> Option<&T> {
-        self.0.get(&entity)
+    fn get(&self, e: Entity) -> Option<&T> {
+        let id = e.get_id();
+        if self.set.contains(id as u32) {
+            let v = unsafe { self.values.0.get_unchecked(id) };
+            if v.generation == e.get_gen() {
+                return Some(&v.data);
+            }
+        }
+        None
+
     }
-    fn get_mut(&mut self, entity: Entity) -> Option<&mut T> {
-        self.0.get_mut(&entity)
+    fn get_mut(&mut self, e: Entity) -> Option<&mut T> {
+        let id = e.get_id();
+        if self.set.contains(id as u32) {
+            let v = unsafe { self.values.0.get_unchecked_mut(id) };
+            if v.generation == e.get_gen() {
+                return Some(&mut v.data);
+            }
+        }
+        None
     }
-    fn insert(&mut self, entity: Entity, value: T) {
-        self.0.insert(entity, value);
+    fn insert(&mut self, e: Entity, mut v: T) {
+        use std::{ptr, mem};
+
+        let id = e.get_id();
+        if self.set.contains(id as u32) {
+            let mut data = &mut self.values.0[id];
+            data.generation = e.get_gen();
+            mem::swap(&mut data.data, &mut v);
+            Some(v)
+        } else {
+            self.set.add(id as u32);
+            if self.values.0.len() <= id {
+                self.extend(id);
+            }
+            unsafe {
+                ptr::write(
+                    &mut self.values.0[id],
+                    GenerationData{
+                        generation: e.get_gen(),
+                        data: v
+                    }
+                );
+            }
+            None
+        };
     }
-    fn remove(&mut self, entity: Entity) -> Option<T> {
-        self.0.remove(&entity)
+    fn remove(&mut self, e: Entity) -> Option<T> {
+        use std::ptr;
+
+        let (id, gen) = (e.get_id(), e.get_gen());
+        let gen_matches = self.values.0.get(id)
+            .map(|x| x.generation == gen).unwrap_or(false);
+
+        if gen_matches && self.set.remove(id as u32) {
+            let value = unsafe { ptr::read(&self.values.0[id]) };
+            Some(value.data)
+        } else {
+            None
+        }
+    }
+    fn entities(&self) -> (&BitSet, &InnerVec<T>) {
+        (&self.set, &self.values)
+    }
+    fn entities_mut(&mut self) -> (&BitSet, &mut InnerVec<T>) {
+        (&self.set, &mut self.values)
+    }
+}
+
+impl<T> super::UnprotectedStorage<T> for InnerVec<T> {
+    unsafe fn get(&self, e: u32) -> &T {
+        &self.0[e as usize].data
+    }
+    unsafe fn get_mut(&mut self, e: u32) -> &mut T {
+        &mut self.0[e as usize].data
```

Comment:
and [get_unchecked_mut](http://doc.rust-lang.org/nightly/std/primitive.slice.html#method.get_unchecked_mut) here